### PR TITLE
Player Z-tracking, again

### PR DESCRIPTION
**Pull request by vuonojenmustaturska**
@ninjanomnom 

:cl: Naksu
code: Preliminary work on tracking cliented living mobs across Z-levels to facilitate mob AI changes later
/:cl:

### CODE REVIEW

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -274,6 +280,17 @@
 			old_area.Exited(src, null)
 		loc = null
 
+/atom/movable/proc/onTransitZ(old_z,new_z)
+	for (var/item in contents) // Notify contents of Z-transition. This can be overridden IF we know the items contents do not care.
```
Comment (Cyberboss) [inline code/game/atoms_movable.dm]: in src

File: code/game/atoms_movable.dm [hunk 2]
```diff
@@ -274,6 +280,17 @@
...
 
+/atom/movable/proc/onTransitZ(old_z,new_z)
+	for (var/item in contents) // Notify contents of Z-transition. This can be overridden IF we know the items contents do not care.
+		var/atom/movable/AM = item
+		AM.onTransitZ(old_z,new_z)
+
+/atom/movable/lighting_object/onTransitZ()
+	return
+
+/atom/movable/light/onTransitZ()
```
Comment (Cyberboss) [inline code/game/atoms_movable.dm]: Should be in the lighting module

File: code/game/turfs/space/space.dm [hunk 3]
```diff
@@ -130,9 +130,11 @@
 		return
 
 	if(destination_z)
+		var/old_z = A.z
 		A.x = destination_x
 		A.y = destination_y
 		A.z = destination_z
+		A.onTransitZ(old_z, destination_z)
```
Comment (Cyberboss) [inline code/game/turfs/space/space.dm]: only if old_z and destz are different

File: code/modules/shuttle/on_move.dm [hunk 4]
```diff
@@ -91,6 +91,8 @@ All ShuttleMove procs go here
 			return
 
 	loc = newT
+	if (newT.z != oldT.z && !istype(moving_dock, /obj/docking_port/mobile/supply)) // Cargo shuttles cannot carry living mobs which are, for now, the only things actually interested in Z transitions
```
Comment (Cyberboss) [inline code/modules/shuttle/on_move.dm]: i'd drop the istype, too snowflake imo

**Review by Cyberboss — DISMISSED** [review]
Why are several onTransitZ's overridden to return immediately?

File: code/modules/shuttle/on_move.dm (hunk 4 above)
Comment (vuonojenmustaturska) [reply code/modules/shuttle/on_move.dm]: I'll leave it until I get to talk with ninjanomnom about shuttle movement in general

**Comment by vuonojenmustaturska** [comment]
The no-op overrides are to prevent the event proc from recursing into the container objects' contents. Ore boxes and vending machines are found on shuttles and cannot contain a mob, but may contain hundreds of if not thousands of atoms.

File: code/modules/shuttle/on_move.dm (hunk 4 above)
Comment (ninjanomnom) [reply code/modules/shuttle/on_move.dm]: the istype is bad here yeah

File: code/modules/shuttle/on_move.dm [hunk 5]
```diff
@@ -91,6 +91,8 @@ All ShuttleMove procs go here
...
+		onTransitZ(oldT.z, newT.z)
```
Comment (ninjanomnom) [inline code/modules/shuttle/on_move.dm]: Anything not crucial to the movement should be done in an after or before proc. The onMove procs are the only part of the move that isn't check ticked. This means the registered z will be inaccurate for a short time but that's fine.

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by vuonojenmustaturska** [comment]
waiting for #33312 to get merged

File: code/modules/mob/living/carbon/human/interactive.dm [hunk 6]
```diff
@@ -828,7 +828,7 @@
 		return pick(/area/hallway, /area/crew_quarters/locker)
 
 /mob/living/carbon/human/interactive/proc/target_filter(target)
-	var/list/filtered_targets = list(/area, /turf, /obj/machinery/door, /atom/movable/light, /obj/structure/cable, /obj/machinery/atmospherics)
+	var/list/filtered_targets = list(/area, /turf, /obj/machinery/door, /atom/movable/lighting_object, /obj/structure/cable, /obj/machinery/atmospherics)
```
[the remark this answers is not in the document]
Comment (duncathan) [reply code/modules/mob/living/carbon/human/interactive.dm]: LOL

File: code/game/machinery/vending.dm [hunk 7]
```diff
@@ -1192,3 +1192,6 @@ IF YOU MODIFY THE PRODUCTS LIST OF A MACHINE, MAKE SURE TO UPDATE ITS RESUPPLY C
 #undef STANDARD_CHARGE
 #undef CONTRABAND_CHARGE
 #undef COIN_CHARGE
+
+/obj/machinery/vending/onTransitZ()
+	return
```
Comment (ninjanomnom) [inline code/game/machinery/vending.dm]: before the undefs

File: code/modules/shuttle/on_move.dm [hunk 8]
```diff
@@ -100,6 +101,11 @@ All ShuttleMove procs go here
 	if(rotation)
 		shuttleRotate(rotation)
 
+	if (src.type != /atom/movable/lighting_object)
+		var/turf/newT = get_turf(src)
+		if (newT.z != oldT.z)
+			onTransitZ(oldT.z, newT.z)
```
Comment (ninjanomnom) [inline code/modules/shuttle/on_move.dm]: I really don't like this solution and would prefer we find something closer to the source of the problem

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/shuttle/on_move.dm (hunk 8 above)
Comment (vuonojenmustaturska) [reply code/modules/shuttle/on_move.dm]: the problem as i understand it is that afterShuttleMove does lighting updates, so some of the moved_atoms will be qdel'd and some might actually become null references depending on how check_tick plays out. Another way to get this working is to check for QDELETED after update_lighting and return, yet another is to never move lighting objects.

**Review by ninjanomnom — DISMISSED** [review]
I'm burning naksu's house down if they don't work on this while the life check is in the code

**Comment by Cyberboss** [comment]
Test merged on bagil

File: code/modules/mob/living/living.dm [hunk 9]
```diff
@@ -1015,3 +1015,27 @@
...
+		if (registered_z)
+			SSmobs.clients_by_zlevel[registered_z] -= src
+		if (client)
+			if (new_z)
+				SSmobs.clients_by_zlevel[new_z] += src
+			registered_z = new_z
+		else
+			registered_z = null
+
+/mob/living/Login()
```
Comment (lzimann) [inline code/modules/mob/living/living.dm]: put this [here](https://github.com/tgstation/tgstation/blob/b9c54f517f7d62516624befcb168cb2ddae3b070/code/modules/mob/living/login.dm#L1)

File: code/modules/mob/living/living.dm [hunk 10]
```diff
@@ -1015,3 +1015,27 @@
...
+	. = ..()
+	var/turf/T = get_turf(src)
+	update_z(T.z)
+
+/mob/living/Logout()
```
Comment (lzimann) [inline code/modules/mob/living/living.dm]: and this one [here](https://github.com/tgstation/tgstation/blob/b9c54f517f7d62516624befcb168cb2ddae3b070/code/modules/mob/living/logout.dm#L1)

**Comment by vuonojenmustaturska** [comment]
I'm going to have to deal with the loc-assigning stuff first due to brainmob-related idiocy, part 1 in #33465

Currently, two things hit the Life() check: brainmobs that Login() in nullspace and are transferred to the real z-level in the same proc without using forceMove or similar (debraining, surgery...), and people on the shuttle because the shuttle transit proc is in the check ticked portion and Life() might tick for some players in between. The latter is not an issue.

**Comment by vuonojenmustaturska** [comment]
Current status: mergeable and usable if you want to, won't spam admins with stupid shit. See above comment for details about current deficiencies. Fixing every loc assignment will be a pretty annoying project.

**Comment by vuonojenmustaturska** [comment]
#33519 once this goes in i'll take another look at this mess

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**